reflex/state.py

```diff
@@ -1065,26 +1315,126 @@ def get_state(self, token: str) -> State:
+                    "or use `@rx.background` decorator for long-running tasks."
+                )
+        await self.redis.set(token, cloudpickle.dumps(state), ex=self.token_expiration)
+
+    @contextlib.asynccontextmanager
+    async def _redis_lock(self, token: str):
+        """Obtain a redis lock for a token.
+
+        Args:
+            token: The token to obtain a lock for.
+
+        Yields:
+            The ID of the lock (to be passed to set_state).
+
+        Raises:
+            RuntimeError: If redis is not configured.
+            LockExpiredError: If the lock has expired while processing the event.
         """
         if self.redis is None:
+            raise RuntimeError("Redis is not configured")
+        lock_key = f"{token}_lock".encode()
+        lock_id = uuid.uuid4().hex.encode()
+        lock_key_channel = f"__keyspace@0__:{lock_key.decode()}"
+
+        def try_get_lock():
+            return self.redis.set(  # pyright: ignore [reportOptionalMemberAccess]
+                lock_key,
+                lock_id,
+                px=self.lock_expiration,
+                nx=True,  # only set if it doesn't exist
+            )
+
+        state_is_locked = await try_get_lock()
+        if not state_is_locked:
+            # Missed the fast-path to get lock, subscribe for lock delete/expire events
+            await self.redis.config_set(
+                "notify-keyspace-events", self._redis_notify_keyspace_events
+            )
+            async with self.redis.pubsub() as pubsub:
+                await pubsub.psubscribe(lock_key_channel)
+                while not state_is_locked:
+                    # wait for the lock to be released
+                    while True:
+                        if not await self.redis.exists(lock_key):
+                            break  # key was removed, try to get the lock again
+                        message = await pubsub.get_message(
+                            ignore_subscribe_messages=True,
+                            timeout=self.lock_expiration / 1000.0,
+                        )
+                        if message is None:
+                            continue
+                        if message["data"] in self._redis_keyspace_lock_release_events:
+                            break
+                    state_is_locked = await try_get_lock()
+        try:
+            yield lock_id
+        except LockExpiredError:
+            state_is_locked = False
+            raise
+        finally:
+            if state_is_locked:
+                # only delete our lock
+                await self.redis.delete(lock_key)
+
+    @contextlib.asynccontextmanager
+    async def modify_state(self, token: str) -> AsyncIterator[State]:
+        """Modify the state for a token while holding exclusive lock.
+
+        Args:
+            token: The token to modify the state for.
+
+        Yields:
+            The state for the token.
+        """
+        if self.redis is not None:
```

Comment:
Later maybe we should make `RedisStateManager` just a subclass of `StateManager` so we don't need to have these special conditional checks, we can just override the methods.